openedx/core/djangoapps/user_authn/api/helper.py

```diff
@@ -84,25 +86,30 @@ def __init__(self, field_type='required'):
                 ):
                     self.valid_fields.append(field_name)
 
+    def _field_exist_in_UserProfile_fields_or_user_extended_config(self, field):
+        return (field in self.userProfile_fields or
+                field in configuration_helpers.get_value('extended_profile_fields', []))
+
     def _get_fields(self):
         """
         Returns the required or optional fields configured in REGISTRATION_EXTRA_FIELDS settings.
         """
         # Custom form fields can be added via the form set in settings.REGISTRATION_EXTENSION_FORM
         custom_form = get_registration_extension_form() or {}
         response = {}
+
         for field in self.valid_fields:
-            if custom_form and field in custom_form.fields:
-                response[field] = form_fields.add_extension_form_field(
-                    field, custom_form, custom_form.fields[field], self.field_type
-                )
-            else:
-                field_handler = getattr(form_fields, f'add_{field}_field', None)
-                if field_handler:
-                    if field == 'confirm_email':
-                        if self.field_type == 'required':
+            if self._field_exist_in_UserProfile_fields_or_user_extended_config(field):
+                if custom_form and field in custom_form.fields:
+                    response[field] = form_fields.add_extension_form_field(
+                        field, custom_form, custom_form.fields[field], self.field_type
+                    )
+                else:
+                    field_handler = getattr(form_fields, f'add_{field}_field', None)
+                    if field_handler:
+                        if field == 'confirm_email' and self.field_type == 'required':
+                            response[field] = field_handler(self.field_type == 'required')
+                        else:
```

Comment:
Rearrangement of code to make it more legible.
```suggestion
            # We don't want "email" field to appear on registration form and "confirm_email"
            # field to appear on welcome/progressive profiling page.
            # Also we don't want to send a field to frontend that cannot be saved at the
            # backend. This is ensured by checking if field can be saved as part of
            # UserProfile model.
            if field == 'confirm_email' and self.field_type == 'optional' or \
                    not self._field_exist_in_UserProfile_fields_or_user_extended_config(field):
                continue

            if custom_form and field in custom_form.fields:
                response[field] = form_fields.add_extension_form_field(
                    field, custom_form, custom_form.fields[field], self.field_type
                )
            else:
                field_handler = getattr(form_fields, f'add_{field}_field', None)
                if field_handler:
                    response[field] = field_handler(self.field_type == 'required')
```